### ISSUE TITLE
502 service unavailable on long requests with nginx

### ISSUE THREAD

**Issue by ehossack** [unbit/uwsgi#2264]
Hi, the bug I'd like to report can be seen with uwsgi and nginx and a long request.

Essentially, during the long request, despite `--socket-timeout` being set to a large value, and nginx's `proxy_read_timeout` being large values, uwsgi seems to have behaviour that causes ngix to drop the connection and return a 502.
```
nginx_1   | 2020/12/04 07:36:21 [error] 19#19: *1 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1, server: localhost:8088, request: "GET /reproduction/ HTTP/1.1", upstream: "http://172.18.0.2:8000/reproduction/", host: "localhost:8088"
```

You can view a reproduction repository here, along with steps to reproduce:

https://github.com/ehossack/potential-uwsgi-bug/blob/mainline/README.md

What I've done is made a simple endpoint that waits 120 seconds before returning a response.
It looks like uWSGI's request consistently is dropped at the 60 second mark, but it actually does complete, but by that point nginx has already given up.

I can confirm that I do not see this behaviour when running the app with apache mod_wsgi or gunicorn.

**Comment by unbit** [comment]
Remove the http router, you are using 2 proxies (nginx and the http router), and the second one (the http router) is tuned to timeout after 60 seconds

**Comment by ehossack** [comment]
Ah, you are saying that the `—http` option is the cause? The solution would then be to run in `server` mode?

From https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html there is no mention of this 60 second timeout.
Can we explain this in the docs? Or is that already noted somewhere (I tried many times to read over the docs)

**Comment by unbit** [comment]
Check the second point of this: https://uwsgi-docs.readthedocs.io/en/latest/ThingsToKnow.html

you are running another layer of proxying without reasons

**Comment by ehossack** [comment]
Definitely, "without reasons". But to me this probably should work (and it appears to for all requests under 60 seconds), however I understand there would be no reason to adjust this since there's a totally acceptable workaround.

In the spirit of making this issue easier to diagnose for the next person, is there a way to document the timeout behaviour? I scanned through the codebase and couldn't see anything that defines where that 60 seconds is set.

**Comment by unbit** [comment]
The option you are searching for is `http-timeout` that sets the timeout of the http router/proxy (you can see it as the equivalent of nginx upstream timeout). At the end you have 3 timeouts in place: nginx (upstream timeout) -> uwsgi-http-router (http-timeout, 60 seconds) -> uwsgi workers (4 seconds, but generally you do not need to touch it unless you do not have a proxy on front).

Relevant option in the code:

https://github.com/unbit/uwsgi/blob/master/plugins/http/http.c#L35

Please make a pull request for this page if you want to add a section for timeouts: https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html 

But please, do not add the http-router if you already have nginx on front, it will hurt performance and will add headaches :)
